File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int result = nums[0], n = nums.size(), l=0, r=0;
        for(int i=0; i<n; i++)
        {
            l = (l ? l : 1)*nums[i];
            r = (r ? r : 1)*nums[n-1-i];
            result = max(result, max(l, r));
        }
        return result;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n=len(nums)
        if n==0:
            raise ValueError
        dp1 = [0]*n
        dp2 = [0]*n
        dp1[0] = nums[0]
        dp2[0] = nums[0]
        for i in range(1, len(nums)):
            dp1[i]=max(dp1[i-1]*nums[i], dp2[i-1]*nums[i], nums[i])
            dp2[i]=min(dp1[i-1]*nums[i], dp2[i-1]*nums[i], nums[i])
        return max(dp1)
        